chcemvediet/apps/accounts/tests/test_forms.py

```diff
@@ -25,23 +25,22 @@ def _create_account_signup_data(self, **kwargs):
                 u'email': u'default_testing_mail@example.com',
                 u'password1': u'default_testing_password',
                 u'password2': u'default_testing_password',
+                u'agreement': True,
                 }
         defaults.update(kwargs)
         return defaults
 
 
-    @unittest.skip(u'FIXME')
     def test_get_signup_form(self):
         response = self.client.get(reverse(u'account_signup'))
         self.assertEqual(response.status_code, 200)
         self.assertTemplateUsed(response, u'account/signup.html')
-        self.assertInHTML(u'<input id="id_first_name" maxlength="30" name="first_name" placeholder="First name" type="text">', response.content)
-        self.assertInHTML(u'<input id="id_last_name" maxlength="30" name="last_name" placeholder="Last name" type="text">', response.content)
-        self.assertInHTML(u'<input id="id_street" maxlength="100" name="street" placeholder="Street" type="text">', response.content)
-        self.assertInHTML(u'<input id="id_city" maxlength="30" name="city" placeholder="City" type="text">', response.content)
-        self.assertInHTML(u'<input id="id_zip" maxlength="5" name="zip" placeholder="Zip" type="text">', response.content)
+        self.assertInHTML(u'<input class="form-control" id="id_first_name" maxlength="30" name="first_name" placeholder="Meno (krstné)" type="text">', unicode(response.content, encoding=u'utf-8'))
```

Comment:
@viliambalaz Nie je dobre, ked testy zavisia na konkretnej lokalizacii. Testy by mali zbehnut bez ohladu na zmeny v lokalizacnych suboroch. Dolezita informacia pre nas, ktoru testujeme, su technicke informacie o fielde, nie zobrazeny label/placeholder. Cize skor zmenit na nieco ako:

```
self.assertRegexpMatches(content, '<input class="form-control" id="id_last_name" maxlength="30" name="last_name" placeholder="[^"]*" type="text" />')
```

Uplne idealne by bolo, keby sa podarilo najst variantu `assertInHTML`, ktora podporuje regexp v hodnotach atributov. Pripadne by podporovala moznost, ze v HTML mozu byt niektore atributy navyse. Tj. aby sme mohli pisat:

```
self.assertFooInHTML(u'<input class="form-control" id="id_first_name" maxlength="30" name="first_name" placeholder=".*" type="text">', content)
```
pripadne:
```
self.assertFooInHTML(u'<input class="form-control" id="id_first_name" maxlength="30" name="first_name" type="text">', content)
```
Skus pogooglit, ci niekde nenajdes kniznicu, kde by taky assert bol? Alebo skusime si ho napisat sami?

Reply:
@martinmacko47 Nasiel som [https://django-with-asserts.readthedocs.io/en/latest/](django-with-asserts) a riesenie by vyzeralo priblizne takto:
```
from with_asserts.mixin import AssertHTMLMixin


class SignupFormTest(AccountsTestCaseMixin, AssertHTMLMixin, TestCase):
        ...
        with self.assertHTML(response, element_id=u'id_first_name') as elem:
            self.assertEqual(elem.attrib[u'name'], u'first_name')
            self.assertIn(u'placeholder', elem.attrib)
            ...
```
alebo druha moznost
```
        self.assertRegexpMatches(content, u'<input class="form-control" id="id_first_name" maxlength="30" name="first_name" placeholder="[^"]*" type="text" \/>')
``` 
Tu si vsak treba davat pozor na whitespace a ordering attributov. Osobne preferujem prvu moznost. Myslim si ze je asi zbytocne komplikovane pisat vlastnu funkcionalitu.

Reply:
@viliambalaz `assertHTML` z `django-with-asserts` vyzera fajn, presne nieco take nam treba. Akurat sa bojim toho, ze autor spravil iba jednu verziu pred osmimi rokmi a od vtedy nic neaktualizoval. Vid repozitar: https://github.com/johnpaulett/django-with-asserts Takze sa obavam, ze ako bude fungovat s novsimi verziami Djanga.

Skus este pohladat, ci nenajdes nieco novsie/popularnesie, co tiez robi inteligentne asserty na HTML. Ak nenajdeme, pouzijeme toto. Ale radsej by som nieco menej stare.

Reply:
@martinmacko47 Moze ostat sposob akym som to napisal, vsetky testovane tagy a ich attributy su zadefinovane v slovniku?

Reply:
Treba teraz tiez spravit kontrolu aj na `self.assertEqual(element.tag, u'input')`

Reply:
@viliambalaz IMHO nie je dobre, ked testy su prilis univerzalne. Lebo ked test padne na riadku `self.assertRegexpMatches`, ze `type` nezodpoveda regexpu `^text$`, tak nebudes vediet, ze o ktory element ide. V beznom kode je to lepsie robit takto univerzalne, ale v testoch nie. Testy by mali byt co najpriamociarejsie ako sa da. Preto by som radsej pisal len:

```
element = html.get_element_by_id(u'id_first_name')
self.assertEqual(element.attrib[u'name'], u'first_name')
self.assertEqual(element.attrib[u'type'], u'text')

element = html.get_element_by_id(u'id_last_name')
self.assertEqual(element.attrib[u'name'], u'last_name')
self.assertEqual(element.attrib[u'type'], u'text')

...
```

Pricom atributy `maxlength` a `placeholder` asi nemusime assertovat. V povodnom teste sa assertovali, lebo `assertInHTML` vynycoval vymenovat vsetky atribudy. Ale tieto dva su v principe v tomto test nepodstatne. Tento test testuje, ci formular ma spravne fieldy. Detaily z fieldov ako dlzka inputu a zobrazeny text su, co sa tyka funkcnosti formularu, nepodstatne.